### PR TITLE
[LLVMCPU][ArmSME] Rework how Arm streaming mode is set on dispatches

### PR DESCRIPTION
**Pull request by MacDue**
Previously, when the `+sme` feature flag was set Armv9 streaming SVE mode would be enabled for all dispatch regions lowered with the following experts:

- `CPUBufferOpsTileAndVectorize`
- `CPUConvTileAndDecomposeExpert`
- `CPUDoubleTilingExpert`

This was not ideal as meant streaming mode could be added to dispatch regions that made no use of scalable vectors, where the (possibly) larger streaming vector length provides no benefit, and there may be a cost due to other overheads.

There was also a flag `--iree-experimental-llvmcpu-arm-force-ssve` which contrary to its name _did not_ force streaming SVE mode. What this flag did do was disable tiling for 2D scalable ArmSME operations, then rely on something else later on setting the streaming mode (but it did not control it).

The patch aims to add clearer and more directed ways to enable streaming mode.

First, streaming mode is no longer set in any lowering experts (it's a fairly low-level concept, that does not need to be configured early in the pipeline). Second, the old `--iree-experimental-llvmcpu-arm-force-ssve` flag is removed.

Now to control tiling for ArmSME and using streaming mode there are two new flags.

`iree-llvmcpu-disable-arm-sme-tiling`:

This disables tiling for ArmSME (i.e. using 2D scalable tile sizes), even when the `+sme` feature flag is set. This results in operations instead being tiled for SVE or Neon (depending on the configuration).

`iree-llvmcpu-force-arm-streaming`:

This enables Arm streaming mode for any dispatch regions that contain scalable vectors. It ignores dispatches that don't contain scalable vectors as enabling streaming mode would provide no benefit.

ci-extra: build_test_all_arm64

### CODE REVIEW

**Comment by MacDue** [comment]
Depends on https://github.com/llvm/llvm-project/pull/94759

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Double negative is confusing. So this is enabling sme tiling by default?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1324]: Echoing earlier comment. My mind is going in circles with double negatives.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:713]: Why is this added twice? Once by default and once behind a flag?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (c-rhodes) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: I think the name could be simplified
```suggestion
        *this, "enable-arm-sme",
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_arm_sme_streaming_mode_tests.mlir
Comment (c-rhodes) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_arm_sme_streaming_mode_tests.mlir]: shall we also check ZA isn't enabled?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_arm_sme_streaming_mode_tests.mlir
Comment (c-rhodes) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_arm_sme_streaming_mode_tests.mlir:38]: I'm not requesting a change but I'm not always convinced the tradeoff between test execution time improvements from split-input-file and readability is worth it. In cases like this where's there a small number of tests I'd rather the attributes were factored out as they add a lot of noise.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy.mlir
Comment (c-rhodes) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy.mlir:4]: I think
```suggestion
// RUN:   --iree-llvmcpu-enable-scalable-vectorization=true --split-input-file --iree-llvmcpu-disable-arm-sme-tiling %s | FileCheck %s --check-prefixes=CHECK,WITH-SSVE
```

would be more accurate?

**Review by c-rhodes — COMMENTED** [review]
thanks for patch Ben. Enabling streaming-mode for the experts you mention was one of our first contributions to IREE, they were chosen simply because one of our pipelines used them, but the intention was always to refine this later. The time has come now this is hurting us.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy.mlir
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy.mlir:4]: It's not SSVE (that decision is not made here). It's simply disabling tiling for ArmSME, whether it uses SSVE or not will be decided much later in the pipeline via the `iree-llvmcpu-force-arm-streaming` flag.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:713]: They are configured differently and enable different things. One enables streaming mode for dispatches that contain scalable vectors (if the `iree-llvmcpu-force-arm-streaming` is set). One enables streaming mode and ZA for dispatches that contain ArmSME ops.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy.mlir
Comment (c-rhodes) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/select_aarch64_sve_lowering_strategy.mlir:4]: ah I see, I missed it was no longer forcing ssve

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I was just following the naming used by other flags, e.g. `iree-llvmcpu-disable-distribution`,  `iree-llvmcpu-disable-vector-peeling`. So with this flag tiling for ArmSME is enabled unless you pass `--iree-llvmcpu-disable-arm-sme-tiling`.

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: My understanding is that if something is opt-in (i.e. not on by default) you have a `enable` flag, and if something is opt-out (i.e. on by default) you have a `disable` flag.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_arm_sme_streaming_mode_tests.mlir
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_arm_sme_streaming_mode_tests.mlir:38]: Keeping this as is, these tests were somewhat of a pain to setup (and I hit some odd errors when I tried to factor things out).

**Review by c-rhodes — APPROVED** [review]
LGTM

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: It would be good to clarify this meaning of this flag a bit. Let me suggest an updated description:
```
Force IREE (and LLVM) to enable the Armv9-A streaming mode for any dispatch region that contains scalable vectors (i.e. switch from SVE to SSVE). Requires the +sme feature flag."
```

Also, add a comment above, e.g.:
```
By default, IREE does not enable the Armv9-A streaming mode in the presence of scalable vectors (even when using `+sme`). This flag will effectively make IREE/LLVM switch from SVE to SSVE in dispatch regions with scalable vectors. 
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: Let's clarify a bit. I think that in situation like this it's good to make it clear that there are separate cases by enumerating :)

```suggestion
              // 1. ArmSME - enable Armv9-A streaming mode (i.e. SSVE) without ZA
              // For dispatch regions that contain scalable vectors (when forced to use streaming mode) and which do not require ZA, enable SSVE without ZA.
```

Feel free to re-use or change, but let's add `1.` and `2.` :) (i.e. same comment for the case below)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]:
```suggestion
          // 2. ArmSME - enable Armv9-A streaming mode (i.e. SSVE) and ZA 
              // Note, dispatch regions with scalable vectors (when forced to use streaming mode) may not need
              // ZA. Hence SSVE and ZA are enabled seperately.
```
```suggestion
              // 1. ArmSME - enable Armv9-A streaming mode (i.e. SSVE) with ZA
              // For dispatch regions that contain 2D scalable vectors, enable both SSVE and ZA.
```

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:95]: I think that we should keep this as experimental and keep the original comment:
> // NOTE: This flag is meant for testing + experimentation and should not be
// used in deployment.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: I also don't quite see how `clDisableArmSMETiling` is a double negative 😅 And yeah, this is similar to e.g. `clDisableDistribution`:
https://github.com/iree-org/iree/blob/5145f7e4e9ab65de1dd91a015440526af678b826/compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L54-L57

**Review by banach-space — APPROVED** [review]
I've made some small suggestions, but overall the logic is sound - LGTM, thanks Ben! (modulo the comments)

@MaheshRavishankar - what do you think?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:95]: It's not entirely clear to me what makes this any more experimental than any other flag? (I note that there's no other `iree-experimental` flags). E.g. compared to `iree-llvmcpu-disable-vector-peeling`, `iree-llvmcpu-disable-distribution`, or `iree-llvmcpu-enable-scalable-vectorization`.

I think this iteration of the flag is much more clearly defined than previously too.

**Review by MaheshRavishankar — COMMENTED** [review]
My main concern here is that there are too many LLVM style flags being added, and I am not sure what combinations of flags is meant to do... Can we have a better story in terms of these flags.

**Comment by MacDue** [comment]
You may not _always_ want to use the ArmSME lowerings (2D scalable outer products, etc) when the `+sme` feature flag is present, so the `--iree-llvmcpu-disable-arm-sme-tiling` flag gives the option to disable tiling for them. Using the ArmSME lowerings  is the default as it's what you'd want in most cases.

SSVE (that is SVE running in streaming mode) is something that needs some consideration for when it should be used. Currently, there's not a proper cost model that can decide when it is beneficial or not. So to allow testing it in the meantime the `--iree-llvmcpu-force-arm-streaming` enables it for any scalable dispatch regions that support it. This is more experimental so it is not enabled by default.

Combination-wise (assuming `+sme,+sve` with scalable vectorization enabled), it works out like this:


|                                      | `--iree-llvmcpu-disable-arm-sme-tiling` | no flag (default)                                            |
|--------------------------------------|-----------------------------------------|------------------------------------------------------|
| **`--iree-llvmcpu-force-arm-streaming`** | Streaming SVE                           | SME (force-arm-streaming is redundant in this case)  |
| **no flag (default)**                            | SVE                                     | SME                                                  |

**Comment by MaheshRavishankar** [comment]
> You may not _always_ want to use the ArmSME lowerings (2D scalable outer products, etc) when the `+sme` feature flag is present, so the `--iree-llvmcpu-disable-arm-sme-tiling` flag gives the option to disable tiling for them. Using the ArmSME lowerings is the default as it's what you'd want in most cases.
> 
> SSVE (that is SVE running in streaming mode) is something that needs some consideration for when it should be used. Currently, there's not a proper cost model that can decide when it is beneficial or not. So to allow testing it in the meantime the `--iree-llvmcpu-force-arm-streaming` enables it for any scalable dispatch regions that support it. This is more experimental so it is not enabled by default.
> 
> Combination-wise (assuming `+sme,+sve` with scalable vectorization enabled), it works out like this:
> 
> `--iree-llvmcpu-disable-arm-sme-tiling`	no flag (default)
> **`--iree-llvmcpu-force-arm-streaming`**	Streaming SVE	SME (force-arm-streaming is redundant in this case)
> **no flag (default)**	SVE	SME

Thanks @MacDue  for the info. Shouldnt the no flag x no flag (the quadrant on the bottom right) be None, i.e. neither use SVE or SME. We dont want to accidently trigger SVE/SME because the `/proc/cpuinfo` has that. Or is this an explicit choice that if a hardware has SME IREEs CPU codegen will try to use it. That seems at odds with the level of support there is right now.

**Comment by MacDue** [comment]
> None, i.e. neither use SVE or SME. We dont want to accidently trigger SVE/SME because the /proc/cpuinfo has that. Or is this an explicit choice that if a hardware has SME IREEs CPU codegen will try to use it.

It is an explicit choice that it will try to use SME/SVE if scalable vectorization is enabled (which that diagram assumes). Scalable vectorization is not on by default right now, so the user still has to opt-in for SME/SVE codegen. 

(note the previous post says "assuming `+sme,+sve` with scalable vectorization enabled")

**Comment by MacDue** [comment]
@MaheshRavishankar We would really like to land this soon :)

**Review by MaheshRavishankar — APPROVED** [review]
Ok, I understand whats happening here now. This looks OK to me now.

**Comment by MacDue** [comment]
Could you land the PR? I don't have write access -- thanks!

**Pull request merged**
